From d0e6dac80f9cdd179da65e7a27a5ee0adff1f3b7 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 17 Nov 2022 09:14:53 +0100
Subject: [PATCH 786/985] evl/lock: add might_hard_lock() to force lock class
 registration

might_hard_lock() is the equivalent of the in-band might_lock()
counterpart, usable with hard spinlocks.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/lock.h  | 24 ++++++++++++++++++++++++
 kernel/evl/mutex.c  |  6 ++----
 kernel/evl/poll.c   |  6 ++----
 kernel/evl/thread.c |  6 ++----
 kernel/evl/wait.c   | 14 ++------------
 5 files changed, 32 insertions(+), 24 deletions(-)

diff --git a/include/evl/lock.h b/include/evl/lock.h
index ffa02fd49..34580d1a7 100644
--- a/include/evl/lock.h
+++ b/include/evl/lock.h
@@ -8,6 +8,30 @@
 
 #include <evl/sched.h>
 
+#ifdef CONFIG_LOCKDEP
+/*
+ * might_hard_lock() disables lazy lock class registration for a given
+ * lock, by forcing it immediately from the current in-band stage so
+ * that this is not postponed until the first acquisition which might
+ * happen from the oob stage, which would be unsafe. We need to make
+ * sure that no oob context potentially grabbing the same lock can
+ * preempt us during the lock->unlock sequence might_lock() performs
+ * not to create invalid chains lockdep could observe, hence the
+ * _full() irq disabling.
+ */
+#define might_hard_lock(__lock)				\
+	do {						\
+		unsigned long __flags;			\
+		inband_context_only();			\
+		local_irq_save_full(__flags);		\
+		might_lock(__lock);			\
+		local_irq_restore_full(__flags);	\
+	} while (0)
+#else
+#define might_hard_lock(__lock)				\
+	do { } while (0)
+#endif
+
 /*
  * A (hard) spinlock API which also deals with thread preemption
  * disabling in the EVL core. Such spinlock may be useful when only
diff --git a/kernel/evl/mutex.c b/kernel/evl/mutex.c
index 56b724ecf..0ae4b80d4 100644
--- a/kernel/evl/mutex.c
+++ b/kernel/evl/mutex.c
@@ -13,6 +13,7 @@
 #include <evl/mutex.h>
 #include <evl/monitor.h>
 #include <evl/wait.h>
+#include <evl/lock.h>
 #include <uapi/evl/signal.h>
 #include <trace/events/evl.h>
 
@@ -474,7 +475,6 @@ void __evl_init_mutex(struct evl_mutex *mutex,
 		const char *name)
 {
 	int type = ceiling_ref ? EVL_MUTEX_PP : EVL_MUTEX_PI;
-	unsigned long flags __maybe_unused;
 
 	mutex->fastlock = fastlock;
 	atomic_set(fastlock, EVL_NO_HANDLE);
@@ -491,9 +491,7 @@ void __evl_init_mutex(struct evl_mutex *mutex,
 #ifdef CONFIG_LOCKDEP
 	lockdep_register_key(&mutex->wchan.lock_key);
 	lockdep_set_class_and_name(&mutex->wchan.lock, &mutex->wchan.lock_key, name);
-	local_irq_save(flags);
-	might_lock(&mutex->wchan.lock);
-	local_irq_restore(flags);
+	might_hard_lock(&mutex->wchan.lock);
 #endif
 }
 EXPORT_SYMBOL_GPL(__evl_init_mutex);
diff --git a/kernel/evl/poll.c b/kernel/evl/poll.c
index 7fcacfabc..0328c7759 100644
--- a/kernel/evl/poll.c
+++ b/kernel/evl/poll.c
@@ -18,6 +18,7 @@
 #include <evl/sched.h>
 #include <evl/flag.h>
 #include <evl/mutex.h>
+#include <evl/lock.h>
 #include <evl/uaccess.h>
 
 struct poll_group {
@@ -605,7 +606,6 @@ int wait_events(struct file *filp,
 static int poll_open(struct inode *inode, struct file *filp)
 {
 	struct poll_group *group;
-	unsigned long flags;
 	int ret;
 
 	group = kzalloc(sizeof(*group), GFP_KERNEL);
@@ -623,9 +623,7 @@ static int poll_open(struct inode *inode, struct file *filp)
 	INIT_LIST_HEAD(&group->waiter_list);
 	evl_init_kmutex(&group->item_lock);
 	raw_spin_lock_init(&group->wait_lock);
-	local_irq_save(flags);
-	might_lock(&group->wait_lock); /* see __evl_init_wait(). */
-	local_irq_restore(flags);
+	might_hard_lock(&group->wait_lock); /* see __evl_init_wait(). */
 	filp->private_data = group;
 	stream_open(inode, filp);
 
diff --git a/kernel/evl/thread.c b/kernel/evl/thread.c
index f4eba804e..e01554fd9 100644
--- a/kernel/evl/thread.c
+++ b/kernel/evl/thread.c
@@ -40,6 +40,7 @@
 #include <evl/factory.h>
 #include <evl/observable.h>
 #include <evl/uaccess.h>
+#include <evl/lock.h>
 #include <trace/events/evl.h>
 
 #define EVL_THREAD_CLONE_FLAGS	\
@@ -170,7 +171,6 @@ int evl_init_thread(struct evl_thread *thread,
 {
 	int state = iattr->flags & ~T_SUSP, ret, gravity;
 	cpumask_var_t affinity;
-	unsigned long flags;
 	va_list args;
 
 	inband_context_only();
@@ -261,9 +261,7 @@ int evl_init_thread(struct evl_thread *thread,
 		lockdep_set_class_and_name(&thread->lock, &thread->lock_key,
 					thread->name);
 	}
-	local_irq_save_full(flags);
-	might_lock(&thread->lock);
-	local_irq_restore_full(flags);
+	might_hard_lock(&thread->lock);
 #endif
 
 	trace_evl_init_thread(thread, iattr, ret);
diff --git a/kernel/evl/wait.c b/kernel/evl/wait.c
index 6a56918e2..1da92a360 100644
--- a/kernel/evl/wait.c
+++ b/kernel/evl/wait.c
@@ -5,6 +5,7 @@
  * Copyright (C) 2001, 2019 Philippe Gerum  <rpm@xenomai.org>
  */
 
+#include <evl/lock.h>
 #include <evl/sched.h>
 #include <evl/wait.h>
 #include <evl/thread.h>
@@ -17,8 +18,6 @@ void __evl_init_wait(struct evl_wait_queue *wq,
 		const char *name,
 		struct lock_class_key *lock_key)
 {
-	unsigned long flags __maybe_unused;
-
 	wq->flags = wq_flags;
 	wq->clock = clock;
 	wq->wchan.pi_serial = 0;
@@ -36,16 +35,7 @@ void __evl_init_wait(struct evl_wait_queue *wq,
 		wq->lock_key_addr.addr = lock_key;
 		lockdep_register_key(lock_key);
 		lockdep_set_class_and_name(&wq->wchan.lock, lock_key, name);
-		/*
-		 * might_lock() forces lockdep to pre-register a
-		 * dynamic lock class, instead of waiting lazily for
-		 * the first lock acquisition, which might happen oob
-		 * for us. Since that registration depends on RCU, we
-		 * need to make sure this happens in-band.
-		 */
-		local_irq_save(flags);
-		might_lock(&wq->wchan.lock);
-		local_irq_restore(flags);
+		might_hard_lock(&wq->wchan.lock);
 	}
 #endif
 }
-- 
2.30.2

